README.md
# Utilu Nightly Tester Tools
Utilu Nightly Tester Tools is an addon for displaying the version number in the title bar and for aiding testers of nightly builds of Mozilla apps like Firefox and Thunderbird. Some features are:

* Copy Build ID to Clipboard
* Copy List of Extensions to Clipboard
* Copy about:support to Clipboard
* Open Profile Folder
* Open Pushlog
* Screenshots utility
* Title bar customization
* Crash options
* Extension compatibility fixing

# Install
You can install the latest stable Utilu Nightly Tester Tools from [utilu.com](http://utilu.com/). To install for development, clone the repo:

    git clone git://github.com/Utilu/nightlytt.git

then add a file titled '{8620c15f-30dc-4dba-a131-7c5d20cf4a40}' to the "extensions" directory in your [profile folder](http://kb.mozillazine.org/Profile_folder) with the text:

    ~/nightlytt/

or whatever the path to your nightlytt folder is.

To package the extension into an xpi for distribution simply `ant` and the default dev script/target will be invoked:

    ant

You can override the build number used in the filename:

    ant -Dbuild.number=1

To build for release (no build number in filename):

    ant release

# Development
All bugs and feature requests are filed in the Nightly Tester Tools project at GitHub. You can [view the list of open issues](https://github.com/Utilu/nightlytt/issues), or you can [file a new issue](https://github.com/Utilu/nightlytt/issues/new). Check out [the wiki](https://wiki.mozilla.org/Auto-tools/Projects/NightlyTesterTools) for a list of current and proposed features and feel free to file bugs and submit patches.

This project uses [.editorconfig](http://editorconfig.org/#overview), which sets defaults for the formatting of the code. So enjoy the use of [compatible editor](http://editorconfig.org/#download). Just download and install the corresponding plugin.

# Related Add-ons
* [Add-on Compatibility Reporter](https://addons.mozilla.org/en-US/firefox/addon/add-on-compatibility-reporter/)
* [Screengrab (fix version)](https://addons.mozilla.org/en-US/firefox/addon/screengrab-fix-version/)


History.md
# Utilu Nightly Tester Tools

3.7.1.4 / 2017-06-20
====================

 * Reverted default extension separator from comma to new line, except for input fields because they don't support newlines
 * Added a hidden preference nightly.extensionSeparator to optionally customize the extension separator
 * Marked extension as multi-process compatible
 * Updated maxVersions for applications
 * Minor changes and improvements

3.7.1.3 / 2016-09-23
====================

 * Improved Customize Title bar Variable Application Update Channel
 * Updated maxVersions for applications
 * Minor changes and improvements

3.7.1.2 / 2016-08-31
====================

 * Optimized the order of the Customize Title bar Variables
 * Added Customize Title bar Variable Application Title and Version
 * Added Customize Title bar Variable Application Display Version
 * Added Customize Title bar Variable Application Beta Version (only filled when applicable)
 * Removed Gecko Build Identifier, since it used PlatformBuildID, just like XUL Platform Build Identifier
 * Removed Gecko Version, since it used PlatformVersion, just like XUL Platform Version
 * Removed unused image file
 * Updated maxVersions for applications
 * Fixed indenting issue
 * Add pushlog feature for all branches
 * Use comma between entries when copying the list of extensions
 * Updated development section of readme and about:nightly
 * Changed the text about where bugs and feature requests are filed in the readme file
 * Applied new default style to all changelogs

3.7.1.1 / 2015-09-10
====================

 * Repository: Created a repository for Utilu Nightly Tester Tools
 * Changed extension name to Utilu Nightly Tester Tools
 * Changed extension ID to {8620c15f-30dc-4dba-a131-7c5d20cf4a40}
 * Removed all partially incomplete translations
 * Removed excessive white space before line endings
 * Removed unused variables module
 * Removed references to End-of-life (EOL) extensions
 * Fixed indentation by always using spaces and never using tabs
 * Removed copy to Pastebin functionality which had become non-working due to API changes
 * Replaced non-working copy to Pastebin functionality by copy text to Clipboard functionality
 * Added spaces between the extensions when copying the list of extensions to the Clipboard
 * Removed non-working image hosts
 * Increased the size of the Customize Title bar window
 * Improved default Customized Title bar title
 * Optimized the order of the Customize Title bar Variables
 * Added Customize Title bar Variable Application Version Pretty
 * Added Customize Title bar Variable Application Update Channel
 * Added Customize Title bar Variable Application Update Channel Pretty (Release, ESR, beta, Aurora, Nighty, Default)
 * Added Customize Title bar Variable Application Version and Channel
 * Added Customize Title bar Variable Number of visible tabs
 * Fixed missing vendor in Title bar Customization in Thunderbird
 * Marked Crash me options which do not cause a crash as such
 * Removed the unused CSS files
 * Cleaned and optimized the remaining CSS files
 * Updated and synced feature lists
 * Updated maxVersions for applications
 * Minor changes and improvements

3.7.1 / 2015-05-26
====================

 * The extension is now signed. Mozilla began requiring all extensions to be signed in order for them to be installable in Release and Beta versions of Firefox. Signing is done through addons.mozilla.org (AMO) and has become mandatory for all extensions.

3.7 / 2013-10-29
====================

 * Include sv-SE locale from BabelZilla (#160)
 * Move extension appmenu entry for Firefox into the secondary pane, and place under Addons menu (#157)
 * Change toolbar button's behavior to offer copy to clipboard if no text box is selected (#153)
 * Add "Build Identifier" Toolbar Icon to SeaMonkey's main window (#154)
 * Overlay Thunderbird's AppMenu (#100)

3.6 / 2013-08-06
====================

 * Include zh-CN locales from BabelZilla. (#143)
 * Make dtd entities for Customize Title bar more localization friendly. (#140)
 * Set file extension in filepicker to avoid silent overwrites. (#124)
 * Add basic Seamonkey support to title bar customization. (#31)
 * Fix for illegal characters in MPL 2.0 headers. (#133)

3.5 / 2013-04-25
====================

 * Use CurProcD as a fallback when referencing GRE specific files. (#115)
 * Detect the application version automatically and setup the right compatibility pref. (#52)
 * Favor the new nsIFilePicker::ShowAsync method in saveScreenshot(). (#112)
 * Remove usage of nsISupportsArray due to its deprecation. (#121)
 * Use getAddonsByTypes() to only retrieve extensions for getExtensionList(). (#94)

3.4 / 2013-02-19
====================

 * Use GreD instead CurProcD to reference GRE specific files. (#115)
 * Update checkCompatibility preferences for compatibility. (#103)
 * Add privacy context to saveScreenshot() due to Bug 795065. (#99)

extension/chrome/content/messenger.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var nightlyApp = {

savedTabmailSetDocumentTitle: null,
tabmailSetDocumentTitle: null,

get defaultTitle() {
  var tabmail = document.getElementById("tabmail");
  return nightlyApp.getWindowTitleForMessenger(tabmail.currentTabInfo);
},

get tabsCount() {
  var tabmail = document.getElementById("tabmail");
  return tabmail.tabInfo.length;
},
get tabTitle() {
  var tabmail = document.getElementById("tabmail");
  return tabmail.currentTabInfo.title;
},

init: function()
{
  var brandbundle = document.getElementById("bundle_brand");
  if (nightly.variables.name==null)
  {
    nightly.variables.name=brandbundle.getString("brandShortName");
  }
  nightly.variables.defaulttitle=brandbundle.getString("brandShortName");
  nightly.variables.brandname=brandbundle.getString("brandFullName");

  var tabmail = document.getElementById("tabmail");
  if (tabmail && typeof(tabmail.setDocumentTitle) === "function")
  {
    nightlyApp.savedTabmailSetDocumentTitle=tabmail.setDocumentTitle;
  }
},

openURL: function(url, event)
{
  var uri = Components.classes["@mozilla.org/network/io-service;1"]
                      .getService(Components.interfaces.nsIIOService)
                      .newURI(url, null, null);

  var protocolSvc = Components.classes["@mozilla.org/uriloader/external-protocol-service;1"]
                              .getService(Components.interfaces.nsIExternalProtocolService);
  protocolSvc.loadUrl(uri);
},

customTabmailSetDocumentTitle: function(aTab)
{
  document.title = nightly.generateText(nightly.getTemplate("titlebar")) + ' - Utilu';
},

updateTitle: function()
{
  if (nightlyApp.savedTabmailSetDocumentTitle) {
    var tabmail = document.getElementById("tabmail");
    if (tabmail.setDocumentTitle != nightlyApp.tabmailSetDocumentTitle)
      tabmail.setDocumentTitle = nightlyApp.tabmailSetDocumentTitle;

    nightlyApp.tabmailSetDocumentTitle(window.document.getElementById("tabmail").currentTabInfo);
  }
},

setCustomTitle: function(title)
{
  nightlyApp.tabmailSetDocumentTitle=nightlyApp.customTabmailSetDocumentTitle;
  nightlyApp.updateTitle();
},

setStandardTitle: function()
{
  nightlyApp.tabmailSetDocumentTitle=nightlyApp.savedTabmailSetDocumentTitle;
  nightlyApp.updateTitle();
},

/**
 * Calculates the title like tabmail's setDocumentTitle(), but doesn't set.
 * See Bug 502389 for the details.
 */
getWindowTitleForMessenger: function(aTab)
{
  let docTitle = aTab.title;

  if (docTitle && !Application.platformIsMac)
    docTitle += document.documentElement
                        .getAttribute("titlemenuseparator");

  if (!docTitle || !Application.platformIsMac)
    docTitle += document.documentElement
                        .getAttribute("titlemodifier");

  return docTitle;
},

openAboutNightly: function () {
  var tabmail = document.getElementById("tabmail");
  tabmail.openTab("contentTab", { contentPage: "about:nightly" });
},

}


extension/chrome/content/titlebar/customize.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

function arrayBasedTreeView(treeViewData) {
  this.data = treeViewData;
}

arrayBasedTreeView.prototype = {
  data: [],

  get rowCount() {
    return this.data.length;
  },

  getCellText: function (aRow, aColumn) {
    return this.data[aRow][aColumn.id];
  },

  setTree: function (aTreebox) {
    this.treebox = aTreebox;
  },

  isContainer: function (aRow) {
    return false;
  },

  isSeparator: function (aRow) {
    return false;
  },

  isSorted: function () {
    return false;
  },

  getLevel: function (aRow) {
    return 0;
  },

  getImageSrc: function (aRow, aCol) {
    return null;
  },

  getRowProperties: function (aRow, aProps) {
    return "";
  },

  getCellProperties: function (aRow, aCol, aProps) {
    return "";
  },

  getColumnProperties: function (aColID, aCol, aProps) {
    return "";
  },

  cycleHeader: function (aCol) {
  },
};


var paneTitle = {

bundle: null,
variables: [],

init: function(aEvent)
{
  aEvent.originalTarget.defaultView.removeEventListener("load", paneTitle.init, false);

  var mediator = Components.classes['@mozilla.org/appshell/window-mediator;1']
              .getService(Components.interfaces.nsIWindowMediator);
  var window = mediator.getMostRecentWindow("navigator:browser");
  if (!window)
    window=mediator.getMostRecentWindow("mail:3pane");
  if (!window)
    window=mediator.getMostRecentWindow("calendarMainWindow");
  if (!window)
    window=mediator.getMostRecentWindow("Songbird:Main");
  if (window)
    paneTitle.nightly=window.nightly;

  paneTitle.toggled();

  paneTitle.bundle=document.getElementById("variablesBundle");

  paneTitle.addVariable("TitleVersion");
  paneTitle.addVariable("DefaultTitle");
  paneTitle.addVariable("TabTitle");
  paneTitle.addVariable("AppID");
  paneTitle.addVariable("BrandName");
  paneTitle.addVariable("Vendor");
  paneTitle.addVariable("Name");
  paneTitle.addVariable("Version");
  paneTitle.addVariable("VersionPretty");
  paneTitle.addVariable("DisplayVersion");
  paneTitle.addVariable("Channel");
  paneTitle.addVariable("ChannelPretty");
  paneTitle.addVariable("BetaVersion");
  paneTitle.addVariable("VersionChannel");
  paneTitle.addVariable("PlatformVersion");
  paneTitle.addVariable("PlatformBuildID");
  paneTitle.addVariable("AppBuildID");
  paneTitle.addVariable("Changeset");
  paneTitle.addVariable("UserAgent");
  paneTitle.addVariable("Locale");
  paneTitle.addVariable("OS");
  paneTitle.addVariable("Processor");
  paneTitle.addVariable("Compiler");
  paneTitle.addVariable("Toolkit");
  paneTitle.addVariable("Profile");
  paneTitle.addVariable("TabsCount");

  paneTitle.setupTree();
},

addVariable: function(name)
{
  var text = null;
  try
  {
    var text = paneTitle.bundle.getString("variable."+name+".description");
  } catch (e) { }
  if (text==null)
  {
    text="";
  }
  var value = paneTitle.nightly.getVariable(name);
  paneTitle.variables.push({variable: "${"+name+"}", description: text, value: value});
},

toggled: function()
{
  var checkbox = document.getElementById("enableTitleBar");
  var text = document.getElementById("customTitle");
  text.disabled=!checkbox.checked;
},

setupTree: function () {
  var tree = document.getElementById("variableTree");
  tree.view = new arrayBasedTreeView(paneTitle.variables);
  tree.addEventListener("click", treeOnClickListener, true);
},
}

function treeOnClickListener(aEvent) {
  if (aEvent.originalTarget.tagName === "treechildren") {
    var tree = aEvent.originalTarget.parentNode;
    var tbo = tree.treeBoxObject;

    // get the row, col and child element at the point
    var row = { }, col = { }, child = { };
    tbo.getCellAt(aEvent.clientX, aEvent.clientY, row, col, child);

    // a workaround to skip extraneous clicks
    if (tree.view.selection.currentIndex === row.value) {
      var titlebox = document.getElementById("customTitle");
      var template = titlebox.value + " " + paneTitle.variables[row.value]["variable"];
      titlebox.value = template;
      // manually set pref, pref change isn't triggered if we just set the value
      paneTitle.nightly.preferences.setCharPref("templates.titlebar", template);
    }
  }
}

window.addEventListener("load",paneTitle.init,false);


extension/chrome/content/nightly.js
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var nightly = {

variables: {
  _appInfo: null,
  get appInfo() {
    if (!this._appInfo) {
      this._appInfo = Components.classes["@mozilla.org/xre/app-info;1"]
                                .getService(Components.interfaces.nsIXULAppInfo)
                                .QueryInterface(Components.interfaces.nsIXULRuntime);
    }
    return this._appInfo;
  },

  get titleversion() { return this.tabtitle + ' - ' + this.vendor + ' ' + this.name + ' ' + this.versionchannel + ' (' + this.appbuildid + ')'; },
  get defaulttitle() { return nightlyApp.defaultTitle; },
  get tabtitle() { return nightlyApp.tabTitle; },
  get appid() this.appInfo.ID,
  brandname: null,
  get vendor() {
    // Fix for vendor not being set in Mozilla Thunderbird
    return this.appInfo.name == "Thunderbird" && this.appInfo.vendor == "" ? "Mozilla" : this.appInfo.vendor;
  },
  get name() this.appInfo.name,
  get version() this.appInfo.version,
  get versionpretty() { return nightly.makeVersionPretty(this.version); },
  get displayversion() {
    var ver = this.version;
    try {
      // The import is only required in Mozilla Thunderbird
      Components.utils.import("resource://gre/modules/AppConstants.jsm");
      if (AppConstants.MOZ_APP_VERSION_DISPLAY) { ver = AppConstants.MOZ_APP_VERSION_DISPLAY; }
    }
    catch (e) {}
    return ver;
  },
  get channel() { return nightly.updateChannel(); },
  get channelpretty() { return nightly.updateChannelPretty(); },
  get betaversion() { return nightly.betaVersion(this.version, this.displayversion); },
  get versionchannel() { return nightly.versionAndChannel(this.version, this.displayversion); },
  get platformversion() this.appInfo.platformVersion,
  get platformbuildid() this.appInfo.platformBuildID,
  get appbuildid() this.appInfo.appBuildID,
  get changeset() { return nightly.getChangeset(); },
  get useragent() navigator.userAgent,
  get locale() {
    var registry = Components.classes["@mozilla.org/chrome/chrome-registry;1"]
                             .getService(Components.interfaces.nsIXULChromeRegistry);
    return registry.getSelectedLocale("global");
  },
  get os() this.appInfo.OS,
  get processor() this.appInfo.XPCOMABI.split("-")[0],
  get compiler() this.appInfo.XPCOMABI.split(/-(.*)$/)[1],
  toolkit: "cairo",
  profile: null,
  get tabscount() { return nightlyApp.tabsCount; },
  flags: ""
},

templates: {
},

getString: function(name, format) {
  if (format) {
    return document.getElementById("nightlyBundle").getFormattedString(name, format);
  }
  else {
    return document.getElementById("nightlyBundle").getString(name);
  }
},

preferences: null,

/**
 *  A helper function for nsIPromptService.confirmEx().
 *  Popping up an alert("Hello!") is as simple as nightly.showConfirmEx({text: "Hello!"});
 *
 *  @see https://developer.mozilla.org/docs/XPCOM_Interface_Reference/nsIPromptService#confirmEx%28%29
 *
 *  @param {Object} aOptions
 *  @param {String} aOptions.text
 *  @param {Number} aOptions.buttonFlags
 *  @param {String} aOptions.button0Title
 *  @param {String} aOptions.button1Title
 *  @param {String} aOptions.button2Title
 *
 *  @returns {Number} Index of the button pressed (0..2)
 */
showConfirmEx: function (aOptions) {
  var promptService = Components.classes["@mozilla.org/embedcomp/prompt-service;1"]
                                .getService(Components.interfaces.nsIPromptService);

  var options = aOptions || {};
  var buttonFlags = options.buttonFlags || promptService.BUTTON_TITLE_OK * promptService.BUTTON_POS_0;

  return promptService.confirmEx(null, "Utilu Nightly Tester Tools", options.text,
    buttonFlags, options.button0Title, options.button1Title, options.button2Title,
    null, {});
},

showAlert: function(id, args) {
  var text = nightly.getString(id, args);
  nightly.showConfirmEx({text: text});
},

init: function() {
  window.removeEventListener("load", nightly.init, false);
  var prefs = Components.classes["@mozilla.org/preferences-service;1"]
                        .getService(Components.interfaces.nsIPrefService);
  nightly.preferences = prefs.getBranch("nightly.")
                             .QueryInterface(Components.interfaces.nsIPrefBranchInternal);
  nightly.preferences.addObserver("", nightly, false);

  var profd = Components.classes["@mozilla.org/file/directory_service;1"]
                        .getService(Components.interfaces.nsIProperties)
                        .get("ProfD", Components.interfaces.nsILocalFile);
  var profservice = Components.classes["@mozilla.org/toolkit/profile-service;1"]
                              .getService(Components.interfaces.nsIToolkitProfileService);
  var profiles = profservice.profiles;
  while (profiles.hasMoreElements()) {
    var profile = profiles.getNext().QueryInterface(Components.interfaces.nsIToolkitProfile);
    if (profile.rootDir.path == profd.path) {
      nightly.variables.profile = profile.name;
      break;
    }
  }

  if (!nightly.variables.profile)
    nightly.variables.profile = profd.leafName;

  nightlyApp.init();
  nightly.prefChange("idtitle");

  var changeset = nightly.getChangeset();
  var currChangeset = nightly.preferences.getCharPref("currChangeset");
  if (!currChangeset || changeset != currChangeset) {
    // keep track of previous nightly's changeset for pushlog
    nightly.preferences.setCharPref("prevChangeset", currChangeset);
    nightly.preferences.setCharPref("currChangeset", changeset);
  }
},

unload: function(pref) {
  window.removeEventListener("unload",nightly.unload,false);
  nightly.preferences.removeObserver("",nightly);
},

prefChange: function(pref) {
  if ((pref == "idtitle") || (pref == "templates.titlebar"))
    nightly.updateTitlebar();
},

updateTitlebar: function() {
  if (nightly.preferences.getBoolPref("idtitle")) {
    var title = nightly.getTemplate("titlebar") + ' - Utilu';
    nightlyApp.setCustomTitle(nightly.generateText(title));
  }
  else {
    nightlyApp.setStandardTitle();
  }
},

observe: function(prefBranch, subject, pref) {
  nightly.prefChange(pref);
},

getStoredItem: function(type, name) {
  name = name.toLowerCase();
  var value = null;
  try {
    return nightly.preferences.getCharPref(type+"."+name);
  }
  catch (e) {}

  if (nightly[type].hasOwnProperty(name)) {
    value = nightly[type][name];
    if (value === undefined || value === null) {
      value = nightly.getString("nightly.variables.nullvalue");
    }
    return value;
  }

  return undefined;
},

getVariable: function(name) {
  return nightly.getStoredItem("variables",name);
},

getTemplate: function(name) {
  return nightly.getStoredItem("templates",name);
},

generateText: function(template) {
  var start = 0;
  var pos = template.indexOf("${",start);
  while (pos >= 0) {
    if ((pos == 0) || (template.charAt(pos - 1) != "$")) {
      var endpos = template.indexOf("}", pos + 2);
      if (endpos >= 0) {
        var varname = template.substring(pos+2,endpos);
        var varvalue = nightly.getVariable(varname);
        if (varvalue !== null && varvalue !== undefined) {
          template = template.substring(0, pos) + varvalue +
                     template.substring(endpos + 1, template.length);
          start = pos + varvalue.length;
        }
        else {
          start = pos + 2;
        }
      }
      else {
        start = pos + 2;
      }
    }
    else {
      start = pos + 2;
    }
    pos = template.indexOf("${", start);
  }
  return template;
},

copyText: function(text) {
  var clipboard = Components.classes["@mozilla.org/widget/clipboardhelper;1"].
                                         getService(Components.interfaces.nsIClipboardHelper);
  clipboard.copyString(text);
},

copyTemplate: function(template) {
  nightly.copyText(nightly.generateText(nightly.getTemplate(template)));
},

parseHTML: function(url, callback) {
  var frame = document.getElementById("sample-frame");
  if (!frame)
    frame = document.createElement("iframe");

  frame.setAttribute("id", "sample-frame");
  frame.setAttribute("name", "sample-frame");
  frame.setAttribute("type", "content");
  frame.setAttribute("collapsed", "true");
  document.getElementById("main-window").appendChild(frame);

  frame.addEventListener("load", function (event) {
    var doc = event.originalTarget;
    if (doc.location.href == "about:blank" || doc.defaultView.frameElement)
      return;

    setTimeout(function () { // give enough time for js to populate page
      callback(doc);
    }, 800);
  }, true);
  frame.contentDocument.location.href = url;
},

makeVersionPretty: function(ver) {
  ver = ver.replace('0a2','0.0.0');
  ver = ver.replace('0a1','0.0.0');
  while (ver.match(new RegExp('\\.','g')).length < 3) { ver += '.0'; }
  return ver;
},
updateChannel: function() {
  return Services.prefs.getCharPref("app.update.channel");
},
updateChannelPretty: function() {
  var channel = nightly.updateChannel();
  if (channel == 'release') { channel = 'Release' }
  else if (channel == 'esr') { channel = 'ESR' }
  else if (channel == 'beta') { channel = 'beta' }
  else if (channel == 'aurora') { channel = 'Aurora' }
  else if (channel == 'nightly') { channel = 'Nighty' }
  else if (channel == 'default') { channel = 'Default' }
  else channel = '';
  return channel;
},
betaVersion: function(ver, displayversion) {
  var beta = displayversion.replace(ver + 'b','');
  if (beta == displayversion) { beta = ''; };
  return beta;
},
versionAndChannel: function(ver, displayversion) {
  var channel = nightly.updateChannelPretty();
  var beta = nightly.betaVersion(ver, displayversion);
  if (channel == 'Release' || channel == 'Default') { channel = ''; }
  if (channel == 'beta' && beta != '') { channel += ' ' + beta; }
  ver = nightly.makeVersionPretty(ver);
  if (channel != '') { ver += ' ' + channel; }
  return ver;
},

copyAboutSupport: function() {
  nightly.parseHTML("about:support", function(doc) {
    var contents = doc.getElementById("contents");
    var text = nightlyPPrint.createTextForElement(contents);
    nightly.copyText(text);
  });
},

menuPopup: function(event, menupopup) {
  if (menupopup == event.target) {
    var attext = false;

    var element = document.commandDispatcher.focusedElement;
    if (element) {
      var type = element.localName.toLowerCase();
      attext = ((type == "input") || (type == "textarea"))
    }

    var node=menupopup.firstChild;
    while (node) {
      if (node.id.indexOf("-insert") != -1)
        node.hidden = !attext;
      if (node.id.indexOf("-copy") != -1)
        node.hidden = attext;
      if (node.id == 'nightly-pushlog-lasttocurrent')
        node.disabled = !nightly.preferences.getCharPref("prevChangeset");
      if (node.id == 'nightly-crashme')
        node.hidden = !ctypes.libraryName;
      if (node.id == 'nightly-compatibility')
        node.setAttribute("checked", nightly.preferences.getBoolPref("disableCheckCompatibility"));
      node=node.nextSibling;
    }
  }
},

insertTemplate: function(template) {
  var element = document.commandDispatcher.focusedElement;
  if (element) {
    var type = element.localName.toLowerCase();
    if ((type == "input") || (type == "textarea")) {
      var text = nightly.generateText(nightly.getTemplate(template));
      var newpos = element.selectionStart+text.length;
      var value = element.value;
      element.value = value.substring(0, element.selectionStart) + text +
                      value.substring(element.selectionEnd);
      element.selectionStart = newpos;
      element.selectionEnd = newpos;
      return;
    }
  }

  // no usable element was found
  const psButtonFlags = Components.interfaces.nsIPromptService;
  var promptOptions = {};
  promptOptions.text = nightly.getString("nightly.notextbox.message") + "\n" +
    nightly.getString("nightly.notextbox.clipboardInstead.message");
  promptOptions.buttonFlags = psButtonFlags.BUTTON_POS_0 * psButtonFlags.BUTTON_TITLE_IS_STRING +
    psButtonFlags.BUTTON_POS_1 * psButtonFlags.BUTTON_TITLE_CANCEL;

  promptOptions.button0Title = nightly.getString("nightly.copyButton.message");

  var buttonPressed = nightly.showConfirmEx(promptOptions);
  if (buttonPressed == 0) {
    nightly.copyTemplate(template);
  }
},

insensitiveSort: function(a, b) {
  a = a.toLowerCase();
  b = b.toLowerCase();
  if (a < b)
    return -1
  if (a > b)
    return 1
  // a must be equal to b
  return 0
},

getExtensionList: function(callback) {
  try {
    Components.utils.import("resource://gre/modules/AddonManager.jsm");

    AddonManager.getAddonsByTypes(['extension'], function(addons) {
      if (!addons.length)
        nightly.showAlert("nightly.noextensions.message", []);

      var strings = addons.map(function(addon) {
        return addon.name + " " + addon.version
          + (addon.userDisabled || addon.appDisabled ? " [DISABLED]" : "");
      });
      strings.sort(nightly.insensitiveSort);
      callback(strings);
    });
  } catch(e) {
    // old extension manager API - take out after Firefox 3.6 support dropped
    var em = Components.classes["@mozilla.org/extensions/manager;1"]
                       .getService(Components.interfaces.nsIExtensionManager);

    var items = em.getItemList(Components.interfaces.nsIUpdateItem.TYPE_EXTENSION, {});

    if (items.length == 0) {
      nightly.showAlert("nightly.noextensions.message", []);
      return null;
    }

    var rdfS = Components.classes["@mozilla.org/rdf/rdf-service;1"]
                         .getService(Components.interfaces.nsIRDFService);
    var ds = em.datasource;
    var disabledResource = rdfS.GetResource("http://www.mozilla.org/2004/em-rdf#disabled");
    var isDisabledResource = rdfS.GetResource("http://www.mozilla.org/2004/em-rdf#isDisabled");
    var text = [];
    for (var i = 0; i < items.length; i++) {
      text[i] = items[i].name + " " + items[i].version;
      var source = rdfS.GetResource("urn:mozilla:item:" + items[i].id);
      var disabled = ds.GetTarget(source, disabledResource, true);
      if (!disabled)
        disabled = ds.GetTarget(source, isDisabledResource, true);
      try {
        disabled=disabled.QueryInterface(Components.interfaces.nsIRDFLiteral);
        if (disabled.Value=="true")
          text[i]+=" [DISABLED]";
      }
      catch (e) { }
    }
    text.sort(nightly.insensitiveSort);
    callback(text);
  }
},

insertExtensions: function() {
  var element = document.commandDispatcher.focusedElement;
  if (element) {
    var type = element.localName.toLowerCase();
    if ((type == "input") || (type == "textarea")) {
      // Input fields don't support new lines, so fallback to a comma separated list
      var extensionSeparator = (type == "input") ? ", " : "\n";
      if (nightly.preferences.prefHasUserValue("extensionSeparator")) {
        extensionSeparator = nightly.preferences.getCharPref("extensionSeparator");
      }
      nightly.getExtensionList(function(text) {
        var newpos = element.selectionStart + text.length;
        var value = element.value;
        element.value = value.substring(0, element.selectionStart) + text.join(extensionSeparator) +
                        value.substring(element.selectionEnd);
        element.selectionStart = newpos;
        element.selectionEnd = newpos;
      });
      return;
    }
  }
  nightly.showAlert("nightly.notextbox.message",[]);
},

copyExtensions: function() {
  var extensionSeparator = "\n";
  if (nightly.preferences.prefHasUserValue("extensionSeparator")) {
    extensionSeparator = nightly.preferences.getCharPref("extensionSeparator");
  }
  nightly.getExtensionList(function(text) {
    if (text)
      nightly.copyText(text.join(extensionSeparator));
  });
},

openProfileDir: function() {
  var stream = Components.classes["@mozilla.org/network/file-input-stream;1"]
                         .createInstance(Components.interfaces.nsIFileInputStream);
  var directoryService = Components.classes["@mozilla.org/file/directory_service;1"]
                                   .getService(Components.interfaces.nsIProperties);

  var profile = directoryService.get("ProfD",Components.interfaces.nsILocalFile);
  try {
    profile.reveal();
  }
  catch (ex) {
    var uri = Components.classes["@mozilla.org/network/io-service;1"]
                        .getService(Components.interfaces.nsIIOService)
                        .newFileURI(profile);
    var protocolSvc = Components.classes["@mozilla.org/uriloader/external-protocol-service;1"]
                                .getService(Components.interfaces.nsIExternalProtocolService);
    protocolSvc.loadUrl(uri);
  }
},

alertType: function(type) {
  var directoryService = Components.classes["@mozilla.org/file/directory_service;1"]
                           .getService(Components.interfaces.nsIProperties);

  var dir = directoryService.get(type, Components.interfaces.nsIFile);
  window.alert(dir.path);
},

getScreenshot: function() {
  window.openDialog("chrome://nightly/content/screenshot/screenshot.xul", "_blank", "chrome,all,dialog=no");
},

openCustomize: function() {
  var wm = Components.classes['@mozilla.org/appshell/window-mediator;1']
                     .getService(Components.interfaces.nsIWindowMediator);

  var win = wm.getMostRecentWindow("NightlyTester:Customize");
  if (win) {
    win.focus();
    return;
  }

  var features;
  try {
    var prefservice = Components.classes["@mozilla.org/preferences-service;1"]
                                .getService(Components.interfaces.nsIPrefBranch);
    var instantApply = prefservice.getBoolPref("browser.preferences.instantApply");
    features = "chrome,titlebar,toolbar,centerscreen,resizable" + (instantApply ? ",dialog=no" : ",modal");
  }
  catch (e) {
    features = "chrome,titlebar,toolbar,centerscreen,resizable,modal";
  }
  openDialog("chrome://nightly/content/titlebar/customize.xul", "", features);
},

getAppIniString : function(section, key) {
  var directoryService = Components.classes["@mozilla.org/file/directory_service;1"].
                           getService(Components.interfaces.nsIProperties);
  var inifile = directoryService.get("GreD", Components.interfaces.nsIFile);
  inifile.append("application.ini");

  if (!inifile.exists()) {
    inifile = directoryService.get("CurProcD", Components.interfaces.nsIFile);
    inifile.append("application.ini");
  }

  var iniParser = Components.manager.getClassObjectByContractID(
                    "@mozilla.org/xpcom/ini-parser-factory;1",
                     Components.interfaces.nsIINIParserFactory)
                  .createINIParser(inifile);
  try {
    return iniParser.getString(section, key);
  } catch (e) {
    return undefined;
  }
},

getRepo: function() {
  return nightly.getAppIniString("App", "SourceRepository");
},

getChangeset: function() {
  return nightly.getAppIniString("App", "SourceStamp");
},

openPushlog: function(fromChange, toChange) {
  if (fromChange) {
    var pushlogUrl = nightly.getRepo() + "/pushloghtml?fromchange=" + fromChange;
    if (toChange)
      pushlogUrl += "&tochange=" + toChange;
    nightlyApp.openURL(pushlogUrl);
  }
},

openPushlogToCurrentBuild: function() {
  var prevChangeset = nightly.preferences.getCharPref("prevChangeset");
  var currChangeset = nightly.getChangeset();
  nightly.openPushlog(prevChangeset, currChangeset);
},

openPushlogSinceCurrentBuild: function() {
  var currChangeset = nightly.getChangeset();
  nightly.openPushlog(currChangeset);
},

toggleCompatibility: function() {
  var forceCompat = nightly.preferences.getBoolPref("disableCheckCompatibility");
  if (nightlyApp.openNotification) {
    var obs = Components.classes["@mozilla.org/observer-service;1"]
              .getService(Components.interfaces.nsIObserverService);
    var restartObserver = {
      observe: function (subject, topic, data) {
        obs.removeObserver(restartObserver, "_nttACS");
        var parsedData = JSON.parse(data);
        if (parsedData && parsedData.restart) {
          nightlyApp.openNotification("nightly-compatibility-restart",
            nightly.getString("nightly.restart.message"),
            nightly.getString("nightly.restart.label"),
            nightly.getString("nightly.restart.accesskey"),
            function() { Application.restart(); });
        }
      }
    };
    obs.addObserver(restartObserver, "_nttACS", false);
  }
  nightly.preferences.setBoolPref("disableCheckCompatibility", !forceCompat);
},

}

try { // import ctypes for determining wether to show crashme menu item
  Components.utils.import("resource://gre/modules/ctypes.jsm");
}
catch(e) {}


// register addon pref listeners
try {
  Components.classes["@mozilla.com/nightly/addoncompatibility;1"].createInstance();
}
catch(e) {}


window.addEventListener("load", nightly.init, false);
window.addEventListener("unload", nightly.unload, false);
